mlflow/pyfunc/__init__.py

```diff
@@ -438,26 +492,45 @@ def spark_udf(spark, model_uri, result_type="double"):
 
     def predict(*args):
         model = SparkModelCache.get_or_load(archive_path)
-        schema = {str(i): arg for i, arg in enumerate(args)}
+        input_schema = model.metadata.get_input_schema()
         pdf = None
+
+        def convert_type(series: pandas.Series, type_: np.dtype):
+            if series.dtype == type_:
+                return series
+            return series.astype(type_, errors="ignore")
+
         for x in args:
             if type(x) == pandas.DataFrame:
                 if len(args) != 1:
                     raise Exception("If passing a StructType column, there should be only one "
                                     "input column, but got %d" % len(args))
                 pdf = x
         if pdf is None:
-            # Explicitly pass order of columns to avoid lexicographic ordering (i.e., 10 < 2)
-            columns = [str(i) for i, _ in enumerate(args)]
-            pdf = pandas.DataFrame(schema, columns=columns)
+            if input_schema is None:
+                data = dict([(str(i), x) for i, x in enumerate(args)])
+                # Explicitly pass order of columns to avoid lexicographic ordering (i.e., 10 < 2)
+                pdf = pandas.DataFrame(data=data, columns=[str(i) for i in range(len(args))])
+            else:
+                if len(args) != len(input_schema.columns):
+                    raise MlflowException("Model input is not compatible with model schema. "
+                                          "Expected  {0} input columns, got {1}".format(
+                                            len(input_schema.columns), len(args)))
+                names = input_schema.column_names()
+                types = input_schema.numpy_types()
+                data = dict([(names[i], convert_type(x, types[i])) for i, x in enumerate(args)])
+                pdf = pandas.DataFrame(data=data, columns=names[:len(args)])
```

Comment:
Nit: why do we slice `names` here? Shouldn't it be the same length as `args`?

Reply:
Oh, good catch. Originally I was allowing a subset of the columns but I decided it was better to require exact match. I'll remove the slice.